rl/agent/actor_critic.py

```diff
@@ -0,0 +1,142 @@
+
+    def compile_models(self):
+        self.actor_opt  = SGD(lr=self.learning_rate)
+        self.critic_opt  = SGD(lr=self.learning_rate)
+        self.actor_loss = "mean_squared_error"
+        self.critic_loss = "mean_squared_error"
+        self.actor.model.compile(loss=self.actor_loss, optimizer=self.actor_opt)
+        self.critic.model.compile(loss=self.critic_loss, optimizer=self.critic_opt)
+
+    def add_custom_objective(self):
+        self.actor_loss = self.custom_objective
+        self.actor.model.compile(loss=self.actor_loss, optimizer=self.actor_opt)
+
+    def recompile_model(self, sys_vars):
+        '''
+        Option to change model optimizer settings
+        Currently only used for changing the learning rate
+        Compiling does not affect the model weights
+        '''
+        if self.epi_change_learning_rate is not None:
+            if (sys_vars['epi'] == self.epi_change_learning_rate and
+                    sys_vars['t'] == 0):
+                self.learning_rate = self.learning_rate / 10.0
+                self.actor_opt = SGD(lr=self.learning_rate)
+                self.critic_opt = SGD(lr=self.learning_rate)
+                self.actor.model.compile(loss=self.actor_loss, optimizer=self.actor_opt)
+                self.critic.model.compile(loss=self.critic_loss, optimizer=self.critic_opt)
+                logger.info('Actor and critic models recompiled with new settings: '
+                            'Learning rate: {}'.format(self.learning_rate))
+            return self.actor.model, self.critic.model
+
+    def select_action(self, state):
+        # Assumes policy is appropriate to this agent
+        return self.policy.select_action(state)
+
+    def compute_action_vals(self, minibatch):
+        return self.actor.model.predict(minibatch['states']) 
+
+    def train_an_epoch(self):
+        if not self.actor_loss_set:
+            self.add_custom_objective()
+            self.actor_loss_set = True
+        minibatch = self.memory.rand_minibatch(self.batch_size)
+        curr_state_vals = self.critic.model.predict(minibatch['states'])
+        next_state_vals = self.critic.model.predict(minibatch['next_states'])
+        curr_targets = minibatch['rewards'] .reshape((minibatch['rewards'].shape[0], 1))+ self.gamma * next_state_vals
+        error = curr_targets - curr_state_vals
+        # Broadcast error to match model output dim
+        error = np.broadcast_to(error, (error.shape[0], self.env_spec['action_dim']))
+        # Mask for actions not taken
+        error = np.multiply(error, minibatch['actions'])
+        loss_critic = self.critic.model.train_on_batch(minibatch['states'], curr_targets)
+        loss_actor = self.actor.model.train_on_batch(minibatch['states'], error)
+        return loss_critic + loss_actor
+
+    def save(self, model_path, global_step=None):
+        logger.info('Saving model checkpoint')
+        self.critic.model.save_weights(os.path.join(model_path, "_critic"))
```

Comment:
beware of memleaks, keras doesn't seem to handle files IO reliably. Is there a way to pass around model weights without saving external files? would fix dqn_freeze too